astropy/coordinates/baseframe.py

```diff
@@ -655,6 +636,43 @@ def __init__(self, *args, **kwargs):
             self._rep_cache = dict()
             self._rep_cache[self._data.__class__.__name__, False] = self._data
 
+        for fnm, fdefault in self.get_frame_attr_names().items():
+            # Read-only frame attributes are defined as FrameAttribue
+            # descriptors which are not settable, so set 'real' attributes as
+            # the name prefaced with an underscore.
+
+            if fnm in kwargs:
+                value = kwargs.pop(fnm)
+                setattr(self, '_' + fnm, value)
+                # Check shape is OK.  As a useful side product, getting the
+                # attribute also validates it.
+                value = getattr(self, fnm)
+                if self.has_data and (getattr(value, 'size', 1) > 1 and
```

Comment:
Checking against ```self.shape``` and only checking the attributes if the frame has data leads to some issues I think. For example, it allows one to do the following:

```python
In [5]: times.shape
Out[5]: (4, 3)

In [6]: altaz_frame = coord.AltAz(location=lapalma, obstime=times, 
...                               temperature=[20,20,20]*u. deg_C)

That doesn't make much sense I think, but it's allowed because the frame has no data and therefore has no ```shape```. This leads me to think that this PR should look carefully at the attributes of the ```BaseCoordinateFrame``` like ```shape```, ```ndim``` and ```isscalar```.